src/firefly/js/util/Interp/LinearInterpolator.js

```diff
@@ -0,0 +1,157 @@
+/**
+ * Lijun Zhang
+ * 05/01/2017
+ * The LinearInterpolator interpolates the one dimensional data points to desired length based on Linear Interpolation
+ * algorithm.  The input data array has to be monotonically ascending or descending.  The minimum length of the input array
+ * is 2.  For any give coordinate (or index of the output array), the LinearInterpolator finds the value of the output array
+ * at that index (or coordinate) based on the input data.
+ */
+
+/**
+ * @Desc This method is displaying error message when there are exceptions.
+ * @param message
+ * @constructor
+ */
+function UserException(message) {
+    this.message = message;
+    this.name = 'UserException';
+}
+
+/**
+ * LinearInterpolator class
+ */
+export class LinearInterpolator {
+    constructor(x, y, allowExtrapolation) {
+        this.x=x;
+        this.y=y;
+        this.n = x.length;
+        this.allowExtrapolation=allowExtrapolation;
+        this.checkInvariant();
+    }
+    /**
+     *
+     * @desc Check that the class invariant is satisfied. The x-coordinates have to be monotonically ascending or
+     * descending. If the input array is not strictly monotonically increasing or decreasing, exception is thrown.
+     * If the input array is descending, reverse the array and return.
+     */
+    checkInvariant() {
+        var revX = false;
+        var x0 = this.x[0]; // first value
+        var xn = this.x[ this.n - 1]; // last value
+
+        if ((this.n < 2)) {
+            throw  new UserException('At least 2 points needed');
+        }
+
+        // determine "x" is ascending or descending
+        var max = ((x0 > xn) ? x0 : xn);
+        for (let i = 1; i < this.n; i++) {
+            if (max > x0) {
+                // check if the data is strictly monotonically increasing
+                if (this.x[i] <= this.x[i - 1]) {
+                    const  x1 = i - 1;
+                    throw new UserException(
+                        'Data must be strictly monotonically increasing or decreasing' + '\n    x['
```

Comment:
This should simple be `throw new Error('....');`. Don't define a UserException, by throwing Error you are carrying a stack trace. Same in other places as well.